pkg/types/provider_mercury.go

```diff
@@ -1,21 +1,97 @@
 package types
 
 import (
-	"github.com/smartcontractkit/chainlink-relay/pkg/reportingplugins/mercury"
-	v1 "github.com/smartcontractkit/chainlink-relay/pkg/reportingplugins/mercury/v1"
-	v2 "github.com/smartcontractkit/chainlink-relay/pkg/reportingplugins/mercury/v2"
-	v3 "github.com/smartcontractkit/chainlink-relay/pkg/reportingplugins/mercury/v3"
+	"context"
+	"math/big"
+
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
 )
 
 // MercuryProvider provides components needed for a mercury OCR2 plugin.
 // Mercury requires config tracking but does not transmit on-chain.
 type MercuryProvider interface {
 	PluginProvider
 
-	ReportCodecV1() v1.ReportCodec
-	ReportCodecV2() v2.ReportCodec
-	ReportCodecV3() v3.ReportCodec
-	OnchainConfigCodec() mercury.OnchainConfigCodec
-	MercuryServerFetcher() mercury.MercuryServerFetcher
-	ChainReader() mercury.ChainReader
+	ReportCodecV1() ReportCodec[int64, MercuryFieldsV1]
+	ReportCodecV2() ReportCodec[uint32, MercuryFieldsV2]
+	ReportCodecV3() ReportCodec[uint32, MercuryFieldsV3]
+	OnchainConfigCodec() MercuryOnchainConfigCodec
+	MercuryServerFetcher() MercuryServerFetcher
+	ChainReader() MercuryChainReader
+}
+
+type MercuryOnchainConfig struct {
+	// applies to all values: price, bid and ask
+	Min *big.Int
+	Max *big.Int
+}
+
+type MercuryOnchainConfigCodec interface {
+	Encode(MercuryOnchainConfig) ([]byte, error)
+	Decode([]byte) (MercuryOnchainConfig, error)
+}
+
+type MercuryServerFetcher interface {
+	// FetchInitialMaxFinalizedBlockNumber should fetch the initial max finalized block number
+	FetchInitialMaxFinalizedBlockNumber(context.Context) (*int64, error)
+	LatestPrice(ctx context.Context, feedID [32]byte) (*big.Int, error)
+	LatestTimestamp(context.Context) (int64, error)
+}
+
+type MercuryChainReader interface {
+	// LatestHeads returns an ordered list of the latest specified number of heads
+	LatestHeads(context.Context, int) ([]Head, error)
+}
+
+type Head struct {
+	Number    uint64
+	Hash      []byte
+	Timestamp uint64
+}
+
+type MercuryFieldsV1 struct {
+	Timestamp             uint32
+	BenchmarkPrice        *big.Int
+	Bid                   *big.Int
+	Ask                   *big.Int
+	CurrentBlockNum       int64
+	CurrentBlockHash      []byte
+	ValidFromBlockNum     int64
+	CurrentBlockTimestamp uint64
+}
+
+type MercuryFieldsV2 struct {
+	ValidFromTimestamp uint32
+	Timestamp          uint32
+	NativeFee          *big.Int
+	LinkFee            *big.Int
+	ExpiresAt          uint32
+	BenchmarkPrice     *big.Int
+}
+
+type MercuryFieldsV3 struct {
+	ValidFromTimestamp uint32
+	Timestamp          uint32
+	NativeFee          *big.Int
+	LinkFee            *big.Int
+	ExpiresAt          uint32
+	BenchmarkPrice     *big.Int
```

Comment:
It looks like this would be a valid simplification too :thinking: 
```suggestion
	MercuryFieldsV2
```

Reply:
We could go with that... 🤔

Reply:
I'm not sure it is really helpful and I'm worried about some kind of edge case failing :shrug:

Reply:
That's fair